src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1628 @@
+            {makeResultLocation(i), ResultLocation{getFunction(), i}});
+        }
+      }
+    }
+  }
+
+  // Calls send values to params in their possible targets, and receive
+  // results.
+  void visitCall(Call* curr) {
+    auto* target = getModule()->getFunction(curr->target);
+    handleCall(
+      curr,
+      [&](Index i) {
+        return LocalLocation{target, i, 0};
+      },
+      [&](Index i) {
+        return ResultLocation{target, i};
+      });
+  }
+  void visitCallIndirect(CallIndirect* curr) {
+    // TODO: the table identity could also be used here
+    auto targetType = curr->heapType;
+    handleCall(
+      curr,
+      [&](Index i) {
+        return SignatureParamLocation{targetType, i};
+      },
+      [&](Index i) {
+        return SignatureResultLocation{targetType, i};
+      });
+  }
+  void visitCallRef(CallRef* curr) {
+    auto targetType = curr->target->type;
+    if (targetType != Type::unreachable) {
+      auto heapType = targetType.getHeapType();
+      handleCall(
+        curr,
+        [&](Index i) {
+          return SignatureParamLocation{heapType, i};
+        },
+        [&](Index i) {
+          return SignatureResultLocation{heapType, i};
+        });
+    }
+  }
+
+  // Creates a location for a null of a particular type and adds a root for it.
+  // Such roots are where the default value of an i32 local comes from, or the
+  // value in a ref.null.
+  Location getNullLocation(Type type) {
+    auto location = NullLocation{type};
+    addRoot(location, PossibleContents::literal(Literal::makeZero(type)));
+    return location;
+  }
+
+  // Iterates over a list of children and adds links from them. The target of
+  // those link is created using a function that is passed in, which receives
+  // the index of the child.
+  void handleChildList(ExpressionList& operands,
+                       std::function<Location(Index)> makeTarget) {
+    Index i = 0;
+    for (auto* operand : operands) {
+      // This helper is not used from places that allow a tuple (hence we can
+      // hardcode the index 0 a few lines down).
+      assert(!operand->type.isTuple());
+
+      if (isRelevant(operand->type)) {
+        info.links.push_back({ExpressionLocation{operand, 0}, makeTarget(i)});
+      }
+      i++;
+    }
+  }
+
+  void visitStructNew(StructNew* curr) {
+    if (curr->type == Type::unreachable) {
+      return;
+    }
+    auto type = curr->type.getHeapType();
+    if (curr->isWithDefault()) {
+      // Link the default values to the struct's fields.
+      auto& fields = type.getStruct().fields;
+      for (Index i = 0; i < fields.size(); i++) {
+        info.links.push_back(
+          {getNullLocation(fields[i].type), DataLocation{type, i}});
+      }
+    } else {
+      // Link the operands to the struct's fields.
+      handleChildList(curr->operands, [&](Index i) {
+        return DataLocation{type, i};
+      });
+    }
+    addRoot(curr, PossibleContents::exactType(curr->type));
```

Comment:
Good point, if we wanted to track fields in allocated structs then we'd need to do so dynamically, and include new information that flows around later. Not simple to do, probably... might require expanding the graph. And yeah, depending on how we do that, maybe this would not be a root, somehow, I'm not sure.